src/sqlfluff/core/rules/std/L039.py

```diff
@@ -0,0 +1,64 @@
+"""Implementation of Rule L039."""
+
+from ..base import BaseCrawler, LintFix, LintResult
+from ..doc_decorators import document_fix_compatible
+
+
+@document_fix_compatible
+class Rule_L039(BaseCrawler):
+    """Unnecessary whitespace found.
+
+    | **Anti-pattern**
+
+    .. code-block::
+
+        SELECT
+            a,        b
+        FROM foo
+
+    | **Best practice**
+    | Unless an indent or preceeding a comment, whitespace should
+    be a single space.
+
+    .. code-block::
+
+        SELECT
+            a, b
+        FROM foo
+    """
+
+    def _eval(self, segment, parent_stack, **kwargs):
+        """Unnecessary whitespace."""
+        # For the given segment, lint whitespace directly within it.
+        prev_newline = True
+        prev_whitespace = None
+        for seg in segment.segments:
```

Comment:
Totally off topic for this PR, but as the number of rules in SQLFluff grows, I wonder if we could provide some higher-level abstractions to simplify code like this.

I could almost imagine something like regex or close to it, but operating at the "segment" level rather than the character level. Wouldn't it be cool if you could write rules that look more like a regex search/capture/replace? 🎉

Reply:
Yeah I've been having the same thought. 🤔 Not sure *how* to do it though...